python_modules/dagster/dagster/_grpc/server.py

```diff
@@ -1254,17 +1256,61 @@ def __init__(
         else:
             self.server_process = server_process
 
+        self._wait_on_exit = wait_on_exit
+
+        self._waited = False
+        self._shutdown = False
+
     @property
     def pid(self):
         return self.server_process.pid
 
     def wait(self, timeout=30):
+        self._waited = True
         if self.server_process.poll() is None:
             seven.wait_for_process(self.server_process, timeout=timeout)
 
-    def create_ephemeral_client(self):
-        from dagster._grpc.client import EphemeralDagsterGrpcClient
+    def __enter__(self):
+        return self
+
+    def __exit__(self, _exception_type, _exception_value, _traceback):
+        self.shutdown_server()
+
+        if self._wait_on_exit:
+            self.wait()
+
+    def shutdown_server(self):
+        if not self._shutdown:
+            self._shutdown = True
+            if self.server_process.poll() is None:
+                try:
+                    self.create_client().shutdown_server()
+                except DagsterUserCodeUnreachableError:
+                    pass
+
+    def __del__(self):
+        if not self._shutdown:
+            warnings.warn(
+                "Local GrpcServerProcess is being destroyed without signalling to server that "
+                "it should shut down. This may result in server processes living longer than "
+                "they need to. To fix this, wrap the GrpcServerProcess in a contextmanager or "
+                "call shutdown_server on it."
+            )
+            import traceback
 
-        return EphemeralDagsterGrpcClient(
-            port=self.port, socket=self.socket, server_process=self.server_process
-        )
+        if not self._waited and os.getenv("STRICT_GRPC_SERVER_PROCESS_WAIT"):
+            warnings.warn(
+                "GrpcServerProcess is being destroyed without waiting for the process to "
+                + "fully terminate. This can cause test instability."
+            )
+
+            # Just for pr testing, do not land
+            traceback.print_stack()
+            from dagster._utils import segfault
+
+            segfault()
```

Comment:
i will be sure to take this out

Reply:
is warning right here or should we raise?

Reply:
I don't think raising in _del_ really does anything - it essentially just warns anyway